cmd/kubeadm/app/util/output/output.go

```diff
@@ -0,0 +1,201 @@
+		outputFormat = *kof.OutputFormat
+	}
+
+	if kof.KubeadmTextPrintFlags != nil {
+		if p, err := kof.KubeadmTextPrintFlags.ToPrinter(outputFormat); !genericclioptions.IsNoCompatiblePrinterError(err) {
+			return p, err
+		}
+	}
+
+	if kof.JSONYamlPrintFlags != nil {
+		if p, err := kof.JSONYamlPrintFlags.ToPrinter(outputFormat); !genericclioptions.IsNoCompatiblePrinterError(err) {
+			return NewResourcePrinterWrapper(kof.TypeSetterPrinter.WrapToPrinter(p, err))
+		}
+	}
+
+	if kof.KubeTemplatePrintFlags != nil {
+		if p, err := kof.KubeTemplatePrintFlags.ToPrinter(outputFormat); !genericclioptions.IsNoCompatiblePrinterError(err) {
+			return NewResourcePrinterWrapper(kof.TypeSetterPrinter.WrapToPrinter(p, err))
+		}
+	}
+
+	return nil, genericclioptions.NoCompatiblePrinterError{OutputFormat: kof.OutputFormat, AllowedFormats: kof.AllowedFormats()}
+}
+
+// AddFlags receives a *cobra.Command reference and binds
+// flags related to Kubeadm printing to it
+func (kof *KubeadmOutputFlags) AddFlags(cmd *cobra.Command) {
+	kof.JSONYamlPrintFlags.AddFlags(cmd)
+	kof.KubeTemplatePrintFlags.AddFlags(cmd)
+	cmd.Flags().StringVarP(kof.OutputFormat, "experimental-output", "o", *kof.OutputFormat, fmt.Sprintf("Output format. One of: %s.", strings.Join(kof.AllowedFormats(), "|")))
+}
+
+// WithDefaultOutput sets a default output format if one is not provided through a flag value
+func (kof *KubeadmOutputFlags) WithDefaultOutput(outputFormat string) *KubeadmOutputFlags {
+	kof.OutputFormat = &outputFormat
+	return kof
+}
+
+// WithTypeSetter sets a wrapper than will surround the returned printer with a printer to type resources
+func (kof *KubeadmOutputFlags) WithTypeSetter(scheme *runtime.Scheme) *KubeadmOutputFlags {
+	kof.TypeSetterPrinter = printers.NewTypeSetter(scheme)
+	return kof
+}
+
+// NewKubeadmOutputFlags creates new KubeadmOutputFlags
+func NewKubeadmOutputFlags(kubeadmTextPrintFlags KubeadmTextPrintFlags) *KubeadmOutputFlags {
+	outputFormat := ""
+
+	kof := &KubeadmOutputFlags{
+		OutputFormat: &outputFormat,
+
+		JSONYamlPrintFlags:     genericclioptions.NewJSONYamlPrintFlags(),
+		KubeTemplatePrintFlags: genericclioptions.NewKubeTemplatePrintFlags(),
+		KubeadmTextPrintFlags:  kubeadmTextPrintFlags,
+	}
+
+	// disable deprecated --template option
+	kof.KubeTemplatePrintFlags.TemplateArgument = nil
+
+	return kof
+}
+
+// KubeadmPrinter is a common printing interface in Kubeadm
+type KubeadmPrinter interface {
+	PrintObj(obj runtime.Object, writer io.Writer) error
+	Info(level klog.Level, args ...interface{})
+	Fprintf(writer io.Writer, format string, args ...interface{}) (n int, err error)
+	Printf(format string, args ...interface{}) (n int, err error)
+}
+
+// TextPrinter implements KubeadmPrinter interface for generic text output
+type TextPrinter struct {
+}
+
+// PrintObj is an implementation of ResourcePrinter.PrintObj that prints object
+func (tp TextPrinter) PrintObj(obj runtime.Object, writer io.Writer) error {
```

Comment:
better have the `tp *TextPrinter` signature for all methods.

Reply:
done